### DISCUSSION TITLE
Websocket parameter returned is null

### DISCUSSION THREAD

**Discussion by pallaviwhy** [aws/aws-iot-device-sdk-js-v2#504] [Q&A]
Hi i am using the example browser pub/sub in which i have changed  provider authentication to fit my requirements.
i am running the code example in vs code

import { fromCognitoIdentityPool } from "@aws-sdk/credential-providers";
import { mqtt, iot} from "aws-iot-device-sdk-v2";

async function connect_websocket(provider: any) {
  return new Promise<mqtt.MqttClientConnection>((resolve, reject) => {
    
    let config: any = iot.AwsIotMqttConnectionConfigBuilder.new_builder_for_websocket()
    config.with_clean_session(true)
    config.with_client_id("test-" + Math.floor(Math.random() * 100000000))
    config.with_endpoint(AWS_IOT_ENDPOINT)
    config.with_credential_provider(provider)
    config.with_use_websockets()
    config.with_keep_alive_seconds(30)
    config.build();

    console.log("Config:",config);

    log("Connecting websocket...");
    const client = new mqtt.MqttClient();

    const  connection= client.new_connection(config);
    

    connection.on("connect", (session_present) => {
      resolve(connection);
    });
    connection.on("interrupt", (error) => {
      log(`Connection interrupted: error=${error}`);
    });
    connection.on("resume", (return_code, session_present) => {
      log(`Resumed: rc: ${return_code} existing session: ${session_present}`);
    });
    connection.on("disconnect", () => {
      log("Disconnected");
    });
    connection.on("error", (error) => {
      reject(error);
    });
    connection.connect();
  });
}

async function main() {
  let idToken = sessionStorage.getItem('idToken');
  if (!idToken) {
    log('No idToken found, please sign in.');
    return;
  }

  let COGNITO_ID = "cognito-idp.us-east-1.amazonaws.com/us-east-1_7YKwIrPtO";
  let loginData = {
    [COGNITO_ID]: idToken,
  };

  let provider;
 

  const providerfunc = await fromCognitoIdentityPool({
    clientConfig: { region: AWS_REGION },
    identityPoolId: AWS_COGNITO_IDENTITY_POOL_ID,
    logins: loginData
  });
  
  provider = await providerfunc();

  console.log("Authenticated", provider);

  connect_websocket(provider)
    .then((connection) => {
      connection
        .subscribe(
          "test/topic",
          mqtt.QoS.AtLeastOnce,
          (topic, payload, dup, qos, retain) => {
            const decoder = new TextDecoder("utf8");
            let message = decoder.decode(new Uint8Array(payload));
            log(`Message received: topic=${topic} message=${message}`);
            // Uncomment the following line to see how disconnect behaves.
            // connection.disconnect();
          }
        )
        .then((subscription) => {
          log(`start publish`);
          var msg_count = 0;
          connection.publish(subscription.topic, `Hello world! ${msg_count}`, subscription.qos);
          // The sample will keep publishing the message every minute.
          setInterval(() => {
            msg_count++;
            const msg = `Hello world! ${msg_count}`;
            connection.publish(subscription.topic, msg, subscription.qos);
          }, 60000);
        });
    })
    .catch((reason) => {
      log(`Error while connecting: ${reason}`);
    });
}

Here is part of my code i have shared 

i have logged the config. here all values are present except that "websocket":{} is null

i have intialized aws_iotendpoint and other inputs correctly and have checked the input are correct

**Comment by bretambrose** [comment]
A few comments:

* The type of `config` (MqttConnectionConfig) does not have a `websocket` parameter.
* You're passing a function object as the credentials provider, but that is not what is expected by the builder/MqttConnectionConfig object.  It is expecting an object that implements the browser `auth.CredentialsProvider` interface
* What is the output when run?
* What happens when you try to use our cognito-based browser pub_sub sample (with your identity pool parameters)?

**Reply by pallaviwhy** [reply]
Here the output of provider:
{
    "identityId": "", [HERE identity id is correct]
    "accessKeyId": "...",
    "secretAccessKey": "...",
    "sessionToken": "...",
    "expiration": " "
}


Output of Config:

{
    "params": {
        "client_id": "test-",
        "host_name": "ENDPOINT.iot.us-east-1.amazonaws.com",   [end point as provided as input]
        "socket_options": {
            "type": 0,
            "domain": ,
            "connect_timeout_ms": ,
            "keepalive": ,
            "keep_alive_interval_sec": ,
            "keep_alive_timeout_sec": ,
            "keep_alive_max_failed_probes": 
        },
        "port": ,
        "clean_session": ,
        "keep_alive": ,
        "username": "?SDK=NodeJSv2&Version=",
        "websocket": {},
        "credentials_provider": {
            "identityId": "...",
            "accessKeyId": "...",
            "secretAccessKey": "...",
            "sessionToken": "...",
            "expiration": ""
        }
    }
}



Only websocket is null which is causing problem

**Reply by bretambrose** [reply]
Please don't post valid credentials, even if expired, in public comments.

**Comment by bretambrose** [comment]
websocket isn't null, it's an empty object.  That parameter exists to override parts of the signing process, which is not needed in your case.  The problem is that you're passing in something that is not a valid credentials provider.

**Reply by pallaviwhy** [reply]
Yea why is it returning null object 
then what is it that i need to pass in order to connect to IoT core.
Also i tried searching for  another builder for mqtt i couldn't find any.

i have tried executing browser/pubsub example inorder to connect to IoT
i am getting errors

/aws-iot-device-sdk-js-v2-main/samples/browser/pub_sub/node_modules/aws-crt/dist/native/auth")'. Did you mean 'AwsCredentialsProvider'?

29 export class AWSCognitoCredentialsProvider extends auth.CredentialsProvider{
                                                           ~~~~~~~~~~~~~~~~~~~


Suggest me to connect to IoT with the code i have provided above. If the provider i am passing is wrong then should be errors but there isn't any with the provider.Then why is it not updating web_socket() object
